Don't read full VM File before decompressing

l0rd opened the pull request:
While working on #21592 we figured out that the full VM File was loaded in memory when detecting the file format, but only a few bytes are needed. This commit addresses that.

[NO NEW TESTS NEEDED]

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

l0rd commented:
cc @mtrmac

Luap99 commented on pkg/machine/compression/decompress.go:
```diff
@@ -19,6 +19,7 @@ const (
 	macOs                = "darwin"
 	progressBarPrefix    = "Extracting compressed file"
 	zipExt               = ".zip"
+	magicNumberMaxSize   = 10
```
This seems fine for now but in the long term I guess we should move this definition to the place where we actually check the bytes, c/storage/pkg/archive.

Luap99 reviewed (approved):
LGTM

giuseppe reviewed (approved):
/lgtm
/hold

mtrmac commented on pkg/machine/compression/decompress.go:
```diff
@@ -31,10 +32,7 @@ type decompressor interface {
-	// Only few bytes are read to detect
-	// the compression type
-	compressedFileContent, err := compressedVMFile.Read()
+	compressedFileContent, err := compressedVMFile.Peek(magicNumberMaxSize)
```
Please rename the variable, and the parameter in `newDecompressor`, to not be misleading.

l0rd replied:
Done.

mtrmac commented on pkg/machine/define/vmfile.go:
```diff
@@ -48,6 +49,17 @@ func (m *VMFile) Read() ([]byte, error) {
+		return nil, err
+	}
+	defer f.Close()
+	r := bufio.NewReader(f)
+	return r.Peek(n)
```
Why not just `f.Read(…)`? 

Involving `bufio` here means having to worry about the buffer size (and this code is, AFAICT, incorrect with `n > defaultBufioSize` — where the latter value is not an API and can change at any time).

l0rd replied:
I have replaced that with `f.Read([]byte)`

mtrmac commented on pkg/machine/define/vmfile.go:
```diff
@@ -50,14 +49,15 @@ func (m *VMFile) Read() ([]byte, error) {
-	return r.Peek(n)
+	b := make([]byte, n)
+	_, err = f.Read(b)
+	return b, err
```
This doesn’t correctly handle files shorter than `n` bytes (and may return `io.EOF` with files exactly `n` bytes, which the caller then doesn’t handle)

mtrmac replied:
```go
n, err = f.Read(b)
if err != nil && err != io.EOF { /* fail */ }
return b[:n], nil
```
I think

mtrmac replied:
… something using `io.ReadAtLeast`, actually? I can’t see any Go guarantee that `Read` will not cause short reads.

mtrmac replied:
`io.ReadFull`, even.

It’s fascinating(?) how many iterations I need to get this right.

Luap99 replied:
but would ReadFull even be correct? It may work in practice but consider we have some new compression arg with 100 bytes magic, and then another compressed file that is less than 100 bytes than you would always error with ReadFull which seem incorrect.

Luap99 replied:
ah i see io.ReadFull will return ErrUnexpectedEOF in this case so that should then be ignored.

mtrmac replied:
Yes, `io.ReadFull` + ignore `EOF` + ignore `ErrUnexpectedEOF` (but fail on other errors).

l0rd replied:
This makes sense in general but I am wondering: wouldn't it be better to fail right away if the machine VM file is shorter than or equal to `magicNumberMaxSize`?

mtrmac replied:
Given that the function is now named specifically `ReadMagicNumber`, sure, that works.

But a `ReadFull`, or an equivalent loop, still seems to be required — or, quite possibly, I’m missing some place where Go promises not to return short reads.

l0rd replied:
From stdlib `(os.File).Read(b)` doc:

> Reads up to len(b) bytes from the File and stores them in b.

==> if the file is shorter than n, it still reads it content without returning any error.

> At end of file, Read returns 0, io.EOF.

==> io.EOF is returned only when there is nothing else to read.

Now, I can still handle `io.EOF` specifically when it's returned by `ReadMagicNumber` but I would not ignore it.

mtrmac replied:
- “up to `len(b)`” would be satisfied by returning just 1 byte, not enough for our magic detection.
- It’s _not_ guaranteed that `io.EOF` is returned _only_ with `n == 0`. https://pkg.go.dev/io#Reader explicitly allows the other behavior, and I can’t see anything for `os.File.Read` documented to override that.

Luap99 commented on pkg/machine/define/vmfile.go:
```diff
@@ -48,6 +48,18 @@ func (m *VMFile) Read() ([]byte, error) {
+		return nil, err
+	}
+	defer f.Close()
+	b := make([]byte, n)
+	_, err = f.Read(b)
+	return b, err
```
```suggestion
	b := make([]byte, n)
	n, err = f.Read(b)
	return b[:n], err
```
this should return `b[:n], err` because in theory we might read less than n.

Luap99 reviewed (commented):
Note in general we prefer the commits to be squashed when addressing code review comments.
`Fixup after mtrmac review` isn't really a helpful commit message once merged into main.

l0rd commented:
> Note in general we prefer the commits to be squashed when addressing code review comments. `Fixup after mtrmac review` isn't really a helpful commit message once merged into main.

For confirmation: preferred way is to sqash + push force right? So even if you lose track of what has been updated in the PR, the commit history is ready for main and doesn't need to be squashed.

mtrmac commented:
Yes (or, optionally, you can create semantically-meaningful multiple commits, and maintain that with `git rebase -i`).

GitHub is making a record of force-pushes in the PR; the record is imperfect but typically good enough to use for keeping up in reviews.

The workflow of adding commits on top of the previous version of a PR doesn’t work if the previous version does not compile at all — the “build each commit” test fails; and we have that test to increase the utility of `git bisect` to work. So force pushes need to be an option, in principle.

l0rd commented:
> Yes (or, optionally, you can create semantically-meaningful multiple commits, and maintain that with `git rebase -i`).
> 
> GitHub is making a record of force-pushes in the PR; the record is imperfect but typically good enough to use for keeping up in reviews.
> 
> The workflow of adding commits on top of the previous version of a PR doesn’t work if the previous version does not compile at all — the “build each commit” test fails; and we have that test to increase the utility of `git bisect` to work. So force pushes need to be an option, in principle.

Fair enough

mtrmac commented on pkg/machine/define/vmfile.go:
```diff
@@ -48,6 +49,22 @@ func (m *VMFile) Read() ([]byte, error) {
+	b := make([]byte, n)
+	n, err = io.ReadFull(f, b)
+	if err != nil && err != io.ErrUnexpectedEOF && err != io.EOF {
+		return b[:n], err
```
(Given what the caller does, `return nil, err` would work just as well here.)

mtrmac reviewed (commented):
LGTM.

mheon commented:
```
--> copying "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\podman-machine-daily.x86_64.hyperv.vhdx" to "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\podman_test3454669177\\podman-machine-daily.x86_64.hyperv.vhdx"/nfailed to close destination file "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\podman_test3454669177\\podman-machine-daily.x86_64.hyperv.vhdx": %!q(<nil>)
```

Error seems legitimate

l0rd commented:
> ```
> --> copying "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\podman-machine-daily.x86_64.hyperv.vhdx" to "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\podman_test3454669177\\podman-machine-daily.x86_64.hyperv.vhdx"/nfailed to close destination file "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\podman_test3454669177\\podman-machine-daily.x86_64.hyperv.vhdx": %!q(<nil>)
> ```
> 
> Error seems legitimate

I am not able to reproduce this error. On my windows 11 with Hyper-V `.\bin\windows\podman.exe machine init` completes successfully:

```
Getting image source signatures
Copying blob 79efdce5b126 done   |
Copying config 44136fa355 done   |
Writing manifest to image destination
79efdce5b126608b8ced856442bc4b192b9ff85eec623d790a016d2212040cd9
Extracting compressed file: podman-machine-default-amd64.vhdx: done
Machine init complete
To start your machine run:

        podman machine start
```

mtrmac commented:
I do think the https://github.com/containers/podman/blob/56e0f063bd280318a861d1c7f816160c31fdf7bb/pkg/machine/e2e/machine_test.go#L125 condition is clearly wrong. But, also, AFAICS, it “only” triggers a `Printf`, not (directly?) a test failure.

l0rd commented:
The error message `failed to close destination file` can be found in the CI logs of [other PRs too](https://cirrus-ci.com/task/5104936988442624) so it's not related to this PR. 

I am going to do a rebase to re-trigger the CI.

Luap99 commented on pkg/machine/define/vmfile.go:
```diff
@@ -48,6 +49,22 @@ func (m *VMFile) Read() ([]byte, error) {
+	n, err = io.ReadFull(f, b)
+	if err != nil && err != io.ErrUnexpectedEOF && err != io.EOF {
+		return b[:n], err
+	} else {
+		return b[:n], nil
+	}
```
nit: the else not required since you return above in the if block
But not worth worth to repush over that.

rhatdan commented:
/unhold
